From 21d237f07ef1b7b176883d7f5a143165cc7c3ee0 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Mon, 8 Feb 2016 10:35:37 -0800
Subject: [PATCH 0655/1566] drm: xilinx: fb: Get vblank before waiting

Make sure vblank is enabled before waiting for it. Otherwise,
it times out when the vblank is disabled.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_fb.c |   18 ++++++++++++++----
 1 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_fb.c b/drivers/gpu/drm/xilinx/xilinx_drm_fb.c
index 95758dd..d8128d9 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_fb.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_fb.c
@@ -169,14 +169,24 @@ int
 xilinx_drm_fb_ioctl(struct fb_info *info, unsigned int cmd, unsigned long arg)
 {
 	struct drm_fb_helper *fb_helper = info->par;
-	struct drm_device *dev = fb_helper->dev;
 	unsigned int i;
+	int ret = 0;
 
 	switch (cmd) {
 	case FBIO_WAITFORVSYNC:
-		for (i = 0; i < fb_helper->crtc_count; i++)
-			drm_wait_one_vblank(dev, i);
-		return 0;
+		for (i = 0; i < fb_helper->crtc_count; i++) {
+			struct drm_mode_set *mode_set;
+			struct drm_crtc *crtc;
+
+			mode_set = &fb_helper->crtc_info[i].mode_set;
+			crtc = mode_set->crtc;
+			ret = drm_crtc_vblank_get(crtc);
+			if (ret) {
+				drm_crtc_wait_one_vblank(crtc);
+				drm_crtc_vblank_put(crtc);
+			}
+		}
+		return ret;
 	default:
 		return -ENOTTY;
 	}
-- 
1.7.5.4

